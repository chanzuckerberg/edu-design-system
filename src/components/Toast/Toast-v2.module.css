@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # TOAST
\*------------------------------------*/

.toast {
  display: flex;
  gap: 1rem;
  padding: 1rem;
  align-items: center;

  border: 0.125rem solid;
  border-left: 1rem solid;
  border-radius: calc(var(--eds-theme-border-radius-objects-sm) * 1px);


  &.toast--status-critical {
    color: var(--eds-theme-color-text-utility-critical);
    background-color: var(--eds-theme-color-background-utility-critical-low-emphasis);
  }

  &.toast--status-favorable {
    color: var(--eds-theme-color-text-utility-favorable);
    background-color: var(--eds-theme-color-background-utility-favorable-low-emphasis);
  }
}

.toast__icon {
  flex-shrink: 0;
}

.toast__body {
  flex-grow: 2;
}