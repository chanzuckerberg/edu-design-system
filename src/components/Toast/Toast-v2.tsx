import clsx from 'clsx';
import React from 'react';

import getIconNameFromStatus from '../../util/getIconNameFromStatus';
import type { Status } from '../../util/variant-types';
import { ButtonV2 as Button } from '../Button';
import Icon from '../Icon';
import Text from '../Text';

import styles from './Toast-v2.module.css';

export type ToastProps = {
  // Component API
  /**
   * Additional class names that can be appended to the component, passed in for styling.
   */
  className?: string;
  /**
   * Callback when notification is dismissed. When passed in, renders banner with a close icon in the top right.
   */
  onDismiss?: () => void;
  // Design API
  /**
   * Keyword to characterize the state of the notification
   */
  status?: Extract<Status, 'favorable' | 'critical'>;
  /**
   * The title/heading of the notification
   */
  title: string;
};

/**
 * `import {Toast} from "@chanzuckerberg/eds";`
 *
 * Toasts display brief, temporary notifications. They're meant to be noticed without disrupting a user's experience or requiring an action to be taken.
 */
export const Toast = ({
  className,
  onDismiss,
  status = 'favorable',
  title,
  ...other
}: ToastProps) => {
  const componentClassName = clsx(
    styles['toast'],
    status && styles[`toast--status-${status}`],
    className,
  );
  return (
    <div className={componentClassName} {...other}>
      <Icon
        className={styles['toast__icon']}
        name={getIconNameFromStatus(status)}
        purpose="decorative"
        size="1.875rem"
      />
      <div className={styles['toast__body']}>
        <Text as="span" className={styles['toast__text']} preset="title-md">
          {title}
        </Text>
      </div>
      {onDismiss && (
        <Button
          aria-label="close"
          context="default"
          icon="close"
          iconLayout="icon-only"
          onClick={onDismiss}
          rank="tertiary"
        >
          Close
        </Button>
      )}
    </div>
  );
};
