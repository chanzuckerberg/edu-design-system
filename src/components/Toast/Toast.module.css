/*------------------------------------*\
    # TOAST
\*------------------------------------*/

/**
 * 1) Message of information, success, caution, or warning to the user
 * 2) Container for the toast content, to provide background styling
 * 3) The thick left border operates as a foreground band of color, 
 *    so foreground rather than border is used
 */

 .toast { /* 2 */
  width: 100%;
  max-width: 25rem;

  padding-left: var(--eds-size-4);
  padding-right: var(--eds-size-1);
  
  display: grid;
  gap: var(--eds-size-1);
  grid-template-columns: 1fr min-content;
  align-items: center;
  
  border: var(--eds-border-width-sm);
  border-left-width: var(--eds-border-width-lg); /* 3 */
  border-radius: var(--eds-size-half);

  border-color: var(--border-light-color);
  border-left-color: var(--border-dark-color);
  
  box-shadow: var(--eds-box-shadow-sm);
  --svg-icon-size-default: 1.25rem;
}

/**
 * 1) Message of information, success, caution, or warning to the user
 * 2) Container for the toast content, to provide background styling
 * 3) The thick left border operates as a foreground band of color, 
 *    so foreground rather than border is used
 */
 
.toast--content {
  padding-top: var(--eds-size-2);
  padding-bottom: var(--eds-size-2);

  display: grid;
  gap: var(--eds-size-2);
  grid-template-columns: min-content 1fr;
  align-items: center;
}
