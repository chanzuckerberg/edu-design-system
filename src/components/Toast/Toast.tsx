import clsx from 'clsx';
import React from 'react';
import styles from './Toast.module.css';
import Button from '../Button';
import Icon from '../Icon';

export type Variant = 'success' | 'error';

export type Props = {
  /**
   * Additional class names that can be appended to the component, passed in for styling.
   */
  className?: string;
  /**
   * The child node(s) contains the toast message. Note: the toast message is displayed inside a TextPassage, so children can contain raw HTML
   */
  children: React.ReactNode;
  /**
   * The color of the Toast, based on EDS defined colors. Also determines the icon used.
   * Note that the Icon mapping matches the style of Banners.
   */
  variant: Variant;
  /**
   * Callback when Toast is dismissed.
   */
  onDismiss?: () => void;
};

/**
 * ```ts
 * import {Toast} from "@chanzuckerberg/eds-components";
 * ```
 *
 * A toast used to provide information on the state of the page, usually in response to a
 * user action. Ex: The user updates their profile, and a toast pop-up informs them that the
 * data was successfully saved.
 */
export const Toast = ({
  children,
  className,
  variant,
  onDismiss,
  // Allow for additional attributes such as aria roles
  ...other
}: Props) => {
  const componentClassName = clsx(
    className,
    styles['toast'],
    variant === 'success' && styles['toast--success'],
    variant === 'error' && styles['toast--error'],
  );
  return (
    <div className={componentClassName} {...other}>
      <div className={styles['toast__content']}>
        <Icon
          name={variant === 'success' ? 'check-circle' : 'warning'}
          purpose="informative"
          size="1.5rem"
          title={variant}
        />
        <p className={styles['toast__text']}>{children}</p>
      </div>
      {onDismiss && (
        <Button onClick={onDismiss} status={variant} variant="icon">
          <Icon
            name="close"
            purpose="informative"
            size="2rem"
            title="dismiss message"
          />
        </Button>
      )}
    </div>
  );
};

export default Toast;
