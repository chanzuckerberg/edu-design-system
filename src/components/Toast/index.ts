export { Toast as default } from './Toast';
export { Toast as ToastV2 } from './Toast-v2';
