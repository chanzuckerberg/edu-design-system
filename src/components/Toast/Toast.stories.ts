import type { StoryObj, Meta } from '@storybook/react';
import type { ComponentProps } from 'react';
import { Toast } from './Toast';

export default {
  title: 'Components/Toast',
  component: Toast,
  parameters: {
    badges: ['1.0'],
  },
  argTypes: { onDismiss: { action: 'dismissed' } },
  args: {
    children: "You've got toast!",
  },
} as Meta<Args>;

type Args = ComponentProps<typeof Toast>;

export const Success: StoryObj<Args> = {
  args: {
    variant: 'success',
  },
};

export const Error: StoryObj<Args> = {
  args: {
    variant: 'error',
  },
};

export const NotDismissable: StoryObj<Args> = {
  args: {
    variant: 'success',
    // @ts-expect-error onDismiss is not nullable, but this is needed to remove the arg from
    // storybook's actions addon
    onDismiss: null,
  },
};
