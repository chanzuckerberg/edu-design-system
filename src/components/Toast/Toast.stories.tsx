import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Toast, Props } from './Toast';
import { TextPassage } from '../TextPassage/TextPassage';
import { LayoutLinelengthContainer } from '../LayoutLinelengthContainer/LayoutLinelengthContainer';
import { Heading } from '../Heading/Heading';
import { Button } from '../Button/Button';

export default {
  title: 'Molecules/Messaging/Toast',
  component: Toast,
} as Meta;

const Template: Story<Props> = (args) => (
  <Toast {...args}>You've got toast!</Toast>
);

export const Default = Template.bind({});
Default.args = { iconName: 'information-circle' };

export const Brand = Template.bind({});
Brand.args = {
  variant: 'brand',
  iconName: 'check-circle',
  title: 'This is a title',
};

export const Success = Template.bind({});
Success.args = {
  variant: 'success',
  iconName: 'check-circle',
  title: 'This is a title',
};

export const Warning = Template.bind({});
Warning.args = {
  variant: 'warning',
  iconName: 'exclamation-circle',
  title: 'This is a title',
};

export const Error = Template.bind({});
Error.args = {
  variant: 'error',
  iconName: 'x-circle',
  title: 'This is a title',
};

export const NotDismissible = Template.bind({});
NotDismissible.args = {
  dismissible: false,
  iconName: 'information-circle',
  title: 'This is a title',
};
