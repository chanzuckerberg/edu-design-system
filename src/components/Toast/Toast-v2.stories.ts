import type { StoryObj, Meta } from '@storybook/react';
import type { ComponentProps } from 'react';

import { Toast } from './Toast-v2';

export default {
  title: 'Components/V2/Toast',
  component: Toast,
  parameters: {
    badges: ['intro-1.0', 'current-2.0'],
  },
  argTypes: { onDismiss: { action: 'dismissed' } },
  args: {
    title: "You've got a temporary notification!",
  },
} as Meta<Args>;

type Args = ComponentProps<typeof Toast>;

export const Default: StoryObj<Args> = {};

export const Favorable: StoryObj<Args> = {
  args: {
    status: 'favorable',
  },
};

export const Critical: StoryObj<Args> = {
  args: {
    status: 'critical',
  },
};

export const NotDismissable: StoryObj<Args> = {
  args: {
    ...Default.args,
    onDismiss: undefined,
  },
};
