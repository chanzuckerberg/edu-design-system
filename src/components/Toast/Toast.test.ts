import { generateSnapshots } from '@chanzuckerberg/story-utils';
import type { StoryFile } from '@storybook/testing-react';
import * as stories from './Toast.stories';

describe('<Toast />', () => {
  generateSnapshots(stories as StoryFile);
});
