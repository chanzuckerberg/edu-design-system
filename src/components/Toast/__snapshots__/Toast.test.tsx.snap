// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toast /> Alert story renders snapshot 1`] = `
<div
  class="toast"
>
  <div
    class="toast--content"
  >
    <p>
      You've got toast!
    </p>
     
  </div>
</div>
`;

exports[`<Toast /> NotDismissable story renders snapshot 1`] = `
<div
  class="toast"
>
  <div
    class="toast--content"
  >
    <p>
      You've got toast!
    </p>
     
  </div>
</div>
`;

exports[`<Toast /> Success story renders snapshot 1`] = `
<div
  class="toast"
>
  <div
    class="toast--content"
  >
    <p>
      You've got toast!
    </p>
     
  </div>
</div>
`;
